### PR TITLE
Update transformers metadata

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a job that will auto-update the repository [transformers-metadata](https://huggingface.co/datasets/huggingface/transformers-metadata) each time it's necessary.

### CODE REVIEW

File: .github/workflows/update_metdata.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            v1-metadata-${{ hashFiles('setup.py') }}
+            v1-metadata
+
+      - name: Setup environment
+        run: |
+          pip install git+https://github.com/huggingface/transformers#egg=transformers[dev]
+
+      - name: Update metadata
+        run: |
+          python utils/update_metadata.py --token ${{ secrets.SYLVAIN_HF_TOKEN }} --commit_sha ${{ github.sha }}
```
Comment (sgugger) [inline .github/workflows/update_metdata.yml:35]: This is a token created just for this ;-)

File: utils/update_metadata.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+# Regexes that match TF/Flax/PT model names.
+_re_tf_models = re.compile(r"TF(.*)(?:Model|Encoder|Decoder|ForConditionalGeneration)")
+_re_flax_models = re.compile(r"Flax(.*)(?:Model|Encoder|Decoder|ForConditionalGeneration)")
+# Will match any TF or Flax model too so need to be in an else branch afterthe two previous regexes.
+_re_pt_models = re.compile(r"(.*)(?:Model|Encoder|Decoder|ForConditionalGeneration)")
+
+
+# Fill this with tuples (pipeline_tag, model_mapping, auto_model)
+PIPELINE_TAGS_AND_AUTO_MODELS = [
```
Comment (sgugger) [inline utils/update_metadata.py:49]: This list will need to be updated when we create new pipelines / auto-classes. I will add a script in repo consistency to make sure it's kept up to date (later today or on Monday).

File: .github/workflows/update_metdata.yml (hunk 1 above)
Comment (julien-c) [reply .github/workflows/update_metdata.yml:35]: cc @SBrandeis :)

File: utils/update_metadata.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+        frameworks_table = get_frameworks_table()
+        frameworks_dataset = Dataset.from_pandas(frameworks_table)
+        frameworks_dataset.to_json(os.path.join(tmp_dir, "frameworks.json"))
+
+        tags_dataset = Dataset.from_json(os.path.join(tmp_dir, "pipeline_tags.json"))
+        table = {tags_dataset[i]["model_class"]: (tags_dataset[i]["pipeline_tag"], tags_dataset[i]["auto_class"]) for i in range(len(tags_dataset))}
+        table = update_pipeline_and_auto_class_table(table)
+
+        # Sort the model classes to avoid some nondeterministic updates to create false update commits.
```
Comment (julien-c) [inline utils/update_metadata.py:186]: 👍

File: utils/update_metadata.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    "Split a camelcased `identifier` into words."
+    matches = re.finditer(".+?(?:(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|$)", identifier)
+    return [m.group(0) for m in matches]
+
+
+def get_frameworks_table():
+    """
+    Generates a dataframe containing the supported auto classes for each model type, using the content of the auto
+    modules.
+    """
```
Comment (julien-c) [inline utils/update_metadata.py:81]: unlike for model classes, do we guarantee that we never have `model_type`s that disappear?

**Review by julien-c — APPROVED** [review]
LGTM, let's try to monitor the first few updates!

File: utils/update_metadata.py (hunk 4 above)
Comment (LysandreJik) [reply utils/update_metadata.py:81]: A `model_type` disappearing would be hugely breaking, so should only happen in very extreme cases. I can't think of a situation where we have done it, or where we would do it to an already merged PR.

File: utils/update_metadata.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            processors[t] = "AutoProcessor"
+        elif t in transformers_module.models.auto.tokenization_auto.TOKENIZER_MAPPING_NAMES:
+            processors[t] = "AutoTokenizer"
+        elif t in transformers_module.models.auto.feature_extraction_auto.FEATURE_EXTRACTOR_MAPPING_NAMES:
+            processors[t] = "AutoFeatureExtractor"
+        else:
+            # Default to AutoTokenizer if a model has nothing, for backward compatibility.
+            processors[t] = "AutoTokenizer"
+    
+    data["processor"] = [processors[t] for t in all_models]
```
Comment (LysandreJik) [inline utils/update_metadata.py:134]: This is a nice approach

**Review by LysandreJik — APPROVED** [review]
Nice, clean usage of `Repository` too :)

**Pull request merged**